File: python/604.py
"""Project Euler Problem 604: Convex path in square.

Find the maximum number of lattice points in an axis-aligned NxN square that an
increasing strictly convex function can pass through.

The optimal order to add segments is to add pairs of segments with the smallest
possible sum of x-distance and y-distance: (1, 1), (1, 2), (2, 1), (1, 3),
(3, 1), (1, 4), (2, 3), (3, 2), (4, 1)... there are φ(k) segments with sum k,
which will increase the width of the square by k*φ(k)/2.

After adding these "complete" levels up to but excluding segments with sum k,
we add pairs of segments that sum to k. Finally, if there is a segment with
both x-distance and y-distance that remains in the boundary of the square, we
can add that single segment.
"""

from __future__ import annotations

from math import ceil, gcd, isqrt


def sieve_phi(limit: int) -> list[int]:
    """Euler totient function sieve."""
    phi = list(range(limit + 1))
    for i in range(2, limit + 1):
        if phi[i] == i:  # i is prime
            for j in range(i, limit + 1, i):
                phi[j] -= phi[j] // i
    return phi


def solve() -> int:
    """Solve Problem 604."""
    N = 10**18
    limit = int(3 * (N ** (1 / 3)))
    phi = sieve_phi(limit)

    ans = 1
    width = 0
    k = 2
    while width + k * phi[k] // 2 <= N:
        ans += phi[k]
        width += k * phi[k] // 2
        k += 1

    num_additions = (N - width) // k
    ans += 2 * num_additions
    width += k * num_additions

    # Find last segment
    found = False
    while not found and width + ceil(k / 2) <= N:
        for big in range(ceil(k / 2), N - width + 1):
            if gcd(k, big) == 1:
                ans += 1
                found = True
                break
        k += 1

    return ans


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/604.c
/*
 * Project Euler Problem 604: Convex path in square
 *
 * Find the maximum number of lattice points in an NxN square that an
 * increasing strictly convex function can pass through.
 *
 * We add pairs of coprime segments sorted by sum k, each contributing
 * k*phi(k)/2 to the width. After full levels, add individual pairs,
 * and possibly one more coprime segment.
 */
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

typedef long long ll;
typedef __int128 i128;

static ll gcd_ll(ll a, ll b) {
    while (b) { ll t = b; b = a % b; a = t; }
    return a;
}

int main(void) {
    ll N = 1000000000000000000LL; /* 10^18 */

    /* Sieve phi up to a safe limit. N^(1/3) ~ 10^6, use 3*10^6 */
    int limit = 3000000;
    int *phi = (int*)malloc((limit + 1) * sizeof(int));
    for (int i = 0; i <= limit; i++) phi[i] = i;
    for (int i = 2; i <= limit; i++) {
        if (phi[i] == i) { /* i is prime */
            for (int j = i; j <= limit; j += i)
                phi[j] -= phi[j] / i;
        }
    }

    ll ans = 1;
    ll width = 0;
    int k = 2;

    while (width + (ll)k * phi[k] / 2 <= N) {
        ans += phi[k];
        width += (ll)k * phi[k] / 2;
        k++;
    }

    ll num_additions = (N - width) / k;
    ans += 2 * num_additions;
    width += k * num_additions;

    /* Find last segment */
    int found = 0;
    while (!found && width + (k + 1) / 2 <= N) {
        for (ll big = (k + 1) / 2; big <= N - width; big++) {
            if (gcd_ll(k, big) == 1) {
                ans += 1;
                found = 1;
                break;
            }
        }
        k++;
    }

    printf("%lld\n", ans);

    free(phi);
    return 0;
}
